executor/cook/executor.py

```diff
@@ -0,0 +1,364 @@
+
+
+class CookExecutor(Executor):
+    """
+        This class is responsible for launching the task sent by the scheduler.
+    """
+    def __init__(self, stop_signal, max_message_length, progress_poll_interval_ms, sandbox_location):
+        self.stop_signal = stop_signal
+        self.completed_signal = Event()
+        self.max_message_length = max_message_length
+        self.progress_poll_interval_ms = progress_poll_interval_ms
+        self.sandbox_location = sandbox_location
+
+    def launchTask(self, driver, task):
+        logging.info('Driver {} launching task {}'.format(driver, task))
+        thread = Thread(target=manage_task,
+                        args=(driver, task, self.stop_signal, self.completed_signal, self.max_message_length,
+                              self.progress_poll_interval_ms, self.sandbox_location, 'stdout', 'stderr'))
+        thread.start()
+
+    def has_stopped(self):
+        return self.completed_signal.isSet()
+
+
+def track_progress(driver, task, progress_watcher, max_message_length, progress_complete_event, send_progress_update):
+    """
+        Sends progress updates to the mesos driver until the stop_signal is set
+        :param driver: The mesos driver to use.
+        :param task: The task to execute.
+        :param progress_watcher: The progress watcher which maintains the current progress state
+        :param max_message_length: Allowed max message length to send
+        :param progress_complete_event: Event that triggers completion of progress tracking
+        :param send_progress_update: The function to invoke while sending progress updates
+        :return: Nothing.
+    """
+    try:
+        task_id = get_task_id(task)
+        for current_progress in progress_watcher.retrieve_progress_states():
+            logging.debug('Latest progress: {}'.format(current_progress))
+            send_progress_update(driver, task_id, max_message_length, current_progress)
+    finally:
+        progress_complete_event.set()
+
+
+def launch_progress_tracker(driver, task, max_message_length, progress_watcher, progress_updater):
+    """
+        Launches the threads that track progress and send progress updates to the driver.
+        :return: Nothing
+    """
+    progress_complete_event = Event()
+    progress_update_thread = Thread(target=track_progress,
+                                    args=(driver, task, progress_watcher, max_message_length, progress_complete_event,
+                                          progress_updater.send_progress_update))
+    progress_update_thread.start()
+    return progress_complete_event
+
+
+class ProgressWatcher(object):
+    """
+        This class tails the output from the target file listening for progress messages.
+        The retrieve_progress_states generates all progress messages iteratively.
+    """
+    @staticmethod
+    def match_progress_update(input_string):
+        result = re.findall(r'\^\^\^\^JOB-PROGRESS: (\d*)(?: )?(.*)', input_string)
```

Comment:
I propose either making this configurable or make it something less crazy and then patching over it for specific formats